In [2]:
from tensorflow import keras
(train_input, train_target), (test_input, test_target) =\
keras.datasets.mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


In [3]:
from sklearn.model_selection import train_test_split
train_scaled = train_input / 255.0 #이미지 픽셀 값 0~1 사이로 변환
train_scaled = train_scaled.reshape(-1, 28*28) #28*28 크기의 2차원 배열을 784크기의 1차원 배열로 펼침
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled, train_target, test_size=0.2, random_state = 42)

인공신경망 모델에 층 2개 추가
* 은닉층의 뉴런 개수를 정하는 데는 특별한 기준이 없으며, 몇 개의 뉴런을 두어야 할지 판단하기 위해서는 상당한 경험이 필요하다 -> 은닉층의 뉴런 개수 바꿔가며 정확도 테스트 해보기!

In [22]:
dense1 = keras.layers.Dense(700, activation = 'sigmoid', input_shape=(784,))  #은닉층
dense2 = keras.layers.Dense(10, activation='softmax')

심층 신경망 만들기

In [23]:
model = keras.Sequential([dense1, dense2])
model.summary()

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 700)               549500    
_________________________________________________________________
dense_13 (Dense)             (None, 10)                7010      
Total params: 556,510
Trainable params: 556,510
Non-trainable params: 0
_________________________________________________________________


은닉층의 뉴런 개수를 증가시키면, 총 파라미터 수도 함께 증가함

모델 훈련

In [24]:
model.compile(loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 14s 9ms/step - loss: 0.3545 - accuracy: 0.8966
Epoch 2/5
1500/1500 [==============================] - 14s 9ms/step - loss: 0.1908 - accuracy: 0.9446
Epoch 3/5
1500/1500 [==============================] - 13s 9ms/step - loss: 0.1319 - accuracy: 0.9610
Epoch 4/5
1500/1500 [==============================] - 14s 10ms/step - loss: 0.1000 - accuracy: 0.9708
Epoch 5/5
1500/1500 [==============================] - 14s 10ms/step - loss: 0.0800 - accuracy: 0.9768


은닉층 추가 결과, 이전에 비해 정확도가 상승하였음 (0.92 -> 0.9639)

**은닉층 개수 조절**

* 50개 : 0.9512
*   100개 : 0.9639
*   200개 : 0.9706
* 300개 : 0.9714
* 400개 : 0.9744
* 500개 : 0.9751   (굉장히 오래 걸리기 시작함)
* 600개 : 0.9757
* 700개 : 0.9768


은닉층 개수 결정 : 400개